### PR TITLE
Skip hooks implementing an expected interface, but loaded by a different classloader

### PR DESCRIPTION
**Pull request by jerrinot**
Fixing to #10124

Problem Statement:
the Mule container has Hazelcast 3.6 embedded. A business application deployed to the container has Hazelcast 3.8 inside. Mule can be configured to prefer application classes -> so an application can initiate Hazelcast 3.8 start. However then the service loading mechanism will find hooks from Hazecast 3.6 -> boom.

This PR is trying to prevent it

Goals:

Everything what was working before should work now. Regression in a patch release is a big deal.
Avoid introducing a new flag / configuration option
The idea is simple:
When a hook implements an interface with the expected name, but it still isn't assignable to the interface then we can assume it was loaded by a different classloader and we can skip it.

When the hook does not implement an interface with the expected name then it's likely a bug -> fail-fast.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java [hunk 1]
```diff
@@ -56,14 +58,16 @@ private BuildInfoProvider() {
...
+        BuildInfo buildInfo = readBuildPropertiesClass(BuildProperties.class, null);
+        try {
+            Class<?> enterpriseClass = BuildInfoProvider.class.getClassLoader()
+                    .loadClass("com.hazelcast.instance.EnterpriseBuildProperties");
+            buildInfo = readBuildPropertiesClass(enterpriseClass, buildInfo);
+        } catch (ClassNotFoundException e) {
+            ignore(e);
         }
+
+        Properties jetProperties = loadPropertiesFromResource("jet-runtime.properties");
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java]: Someday, we're going to do the same for Jet, right?

File: hazelcast/src/main/java/com/hazelcast/util/ServiceLoader.java [hunk 2]
```diff
@@ -338,32 +342,106 @@ public void remove() {
...
                 throw new HazelcastException(e);
             }
         }
 
+        private void onNonAssignableClass(String className, Class candidate) {
+            if (expectedType.isInterface()) {
+                if (ClassLoaderUtil.implementsInterfaceWithSameName(candidate, expectedType)) {
+                    // this can happen in application containers - different Hazelcast JARs are loaded
+                    // by different classloaders.
+                    LOGGER.fine("There appears to be a classloading conflict. "
```
Comment (mdogan) [inline hazelcast/src/main/java/com/hazelcast/util/ServiceLoader.java]: Should this logged more visible? like info or warning?

File: hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java (hunk 1 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/instance/BuildInfoProvider.java]: correct. but we can do it only after Jet expose the class.

File: hazelcast/src/main/java/com/hazelcast/util/ServiceLoader.java (hunk 2 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/util/ServiceLoader.java]: I am not sure. it's probably a business as usual when deployed an app. container. 
it provides no value unless you are troubleshooting it.

**Comment by jerrinot** [comment]
run-lab-run

**Pull request merged**